File: backend/app/schemas/community.py
from datetime import datetime
from pydantic import BaseModel, Field, model_validator

POST_TYPES = {"artwork", "advice", "tutorial", "retouch", "equipment", "location", "before_after"}

class ImageInput(BaseModel):
    image_url: str = Field(max_length=500); thumbnail_url: str | None = None; original_url: str | None = None
    sort_order: int = 0; width: int | None = None; height: int | None = None; file_size: int | None = None
    mime_type: str | None = None; image_role: str = "normal"; alt_text: str | None = Field(None, max_length=300)

class PostPayload(BaseModel):
    title: str = Field("", max_length=160); content: str = Field("", max_length=10000)
    post_type: str = "artwork"; visibility: str = "public"; status: str = "draft"
    allow_comments: bool = True; allow_ai_review: bool = True; allow_original_download: bool = False
    location_name: str | None = Field(None, max_length=160); device_name: str | None = Field(None, max_length=120)
    lens_name: str | None = Field(None, max_length=120); aperture: str | None = None; shutter_speed: str | None = None
    iso: int | None = Field(None, ge=1, le=409600); focal_length: str | None = None
    editing_software: str | None = None; editing_notes: str | None = Field(None, max_length=3000)
    images: list[ImageInput] = Field(default_factory=list, max_length=9); tags: list[str] = Field(default_factory=list, max_length=10)
    copyright_confirmed: bool = False
    @model_validator(mode="after")
    def validate_post(self):
        if self.post_type not in POST_TYPES: raise ValueError("不支持的帖子类型")
        if self.visibility not in {"public", "followers", "private"}: raise ValueError("无效可见范围")
        if self.status not in {"draft", "published"}: raise ValueError("无效状态")
        if self.status == "published" and (not self.title.strip() or not self.content.strip() or not self.images or not self.copyright_confirmed): raise ValueError("发布时需填写标题、正文、上传图片并确认版权")
        roles = {i.image_role for i in self.images}
        if self.post_type == "before_after" and self.status == "published" and not {"before", "after"}.issubset(roles): raise ValueError("前后对比帖需包含 before 和 after 图片")
        return self

class PostUpdate(PostPayload):
    copyright_confirmed: bool = True

class CommentPayload(BaseModel):
    content: str = Field(min_length=1, max_length=2000); parent_id: int | None = None; reply_to_user_id: int | None = None

class CollectionPayload(BaseModel):
    name: str = Field(min_length=1, max_length=80); description: str | None = Field(None, max_length=500); visibility: str = "private"; cover_image_url: str | None = None

class FavoritePayload(BaseModel): collection_id: int | None = None
class ReportPayload(BaseModel): target_type: str; target_id: int; reason: str; description: str | None = Field(None, max_length=1000); evidence_urls: list[str] = Field(default_factory=list, max_length=5)

File: frontend/src/api/community.ts
import { apiRequest } from "./client";
import { uploadImage } from "./upload";

export type CommunityImage={id?:number;image_url:string;thumbnail_url?:string|null;sort_order:number;width?:number|null;height?:number|null;image_role:string;alt_text?:string|null};
export type CommunityPost={id:number;author:{id:number;username:string;avatar_url?:string|null;signature?:string|null};title:string;content:string;post_type:string;visibility:string;status:string;allow_comments:boolean;allow_ai_review:boolean;allow_original_download:boolean;location_name?:string|null;device_name?:string|null;lens_name?:string|null;aperture?:string|null;shutter_speed?:string|null;iso?:number|null;focal_length?:string|null;editing_software?:string|null;editing_notes?:string|null;cover_image_url?:string|null;images:CommunityImage[];tags:{name:string;slug:string;category:string}[];view_count:number;like_count:number;favorite_count:number;comment_count:number;share_count:number;published_at?:string|null;created_at:string;updated_at:string;is_liked:boolean;is_favorited:boolean;is_following_author:boolean;is_owner:boolean};
export type PostInput=Omit<Partial<CommunityPost>,"id"|"author"|"tags"|"images">&{title:string;content:string;post_type:string;visibility:string;status:string;images:Omit<CommunityImage,"id">[];tags:string[];copyright_confirmed:boolean};
export type CommunityComment={id:number;content:string;parent_id?:number|null;reply_to_user_id?:number|null;like_count:number;reply_count:number;created_at:string;is_liked:boolean;is_owner:boolean;author:{id:number;username:string;avatar_url?:string|null}};

export const getFeed=(kind="recommended",cursor?:number)=>apiRequest<{items:CommunityPost[];next_cursor?:number|null}>(`/community/feed/${kind}${cursor?`?cursor=${cursor}`:""}`);
export const getPost=(id:number)=>apiRequest<CommunityPost>(`/community/posts/${id}`);
export const createPost=(data:PostInput)=>apiRequest<CommunityPost>("/community/posts",{method:"POST",body:JSON.stringify(data)});
export const updatePost=(id:number,data:PostInput)=>apiRequest<CommunityPost>(`/community/posts/${id}`,{method:"PATCH",body:JSON.stringify(data)});
export const deletePost=(id:number)=>apiRequest<void>(`/community/posts/${id}`,{method:"DELETE"});
export const likePost=(id:number)=>apiRequest<{liked:boolean;like_count:number}>(`/community/posts/${id}/like`,{method:"POST"});
export const unlikePost=(id:number)=>apiRequest<{liked:boolean;like_count:number}>(`/community/posts/${id}/like`,{method:"DELETE"});
export const favoritePost=(id:number)=>apiRequest<{favorited:boolean;favorite_count:number}>(`/community/posts/${id}/favorite`,{method:"POST",body:"{}"});
export const unfavoritePost=(id:number)=>apiRequest<{favorited:boolean;favorite_count:number}>(`/community/posts/${id}/favorite`,{method:"DELETE"});
export const getComments=(id:number)=>apiRequest<CommunityComment[]>(`/community/posts/${id}/comments`);
export const addComment=(id:number,content:string,parent_id?:number)=>apiRequest<{id:number}>(`/community/posts/${id}/comments`,{method:"POST",body:JSON.stringify({content,parent_id})});
export const likeComment=(id:number)=>apiRequest<{liked:boolean;like_count:number}>(`/community/comments/${id}/like`,{method:"POST"});
export const unlikeComment=(id:number)=>apiRequest<{liked:boolean;like_count:number}>(`/community/comments/${id}/like`,{method:"DELETE"});
export const deleteComment=(id:number)=>apiRequest<void>(`/community/comments/${id}`,{method:"DELETE"});
export const uploadCommunityImage=(file:File)=>uploadImage(file,"community");
export const getNotifications=()=>apiRequest<any[]>("/community/notifications");
export const markNotificationsRead=()=>apiRequest("/community/notifications/read-all",{method:"POST"});
